tsl/src/continuous_aggs/create.c

```diff
@@ -1097,38 +1110,66 @@ cagg_validate_query(Query *query, bool finalized)
 	/* check if we have a hypertable in the FROM clause */
 	rtref = linitial_node(RangeTblRef, query->jointree->fromlist);
 	rte = list_nth(query->rtable, rtref->rtindex - 1);
+
 	/* FROM only <tablename> sets rte->inh to false */
-	if (rte->relkind != RELKIND_RELATION || rte->tablesample || rte->inh == false)
+	if ((rte->relkind != RELKIND_RELATION && rte->relkind != RELKIND_VIEW) || rte->tablesample ||
+		rte->inh == false)
 	{
 		ereport(ERROR,
 				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 				 errmsg("invalid continuous aggregate view")));
 	}
-	if (rte->relkind == RELKIND_RELATION)
+
+	if (rte->relkind == RELKIND_RELATION || rte->relkind == RELKIND_VIEW)
 	{
 		const Dimension *part_dimension = NULL;
+		int32 parent_mat_hypertable_id = INVALID_HYPERTABLE_ID;
+
+		if (rte->relkind == RELKIND_RELATION)
+			ht = ts_hypertable_cache_get_cache_and_entry(rte->relid, CACHE_FLAG_NONE, &hcache);
+		else
+		{
+			const ContinuousAgg *cagg;
 
-		ht = ts_hypertable_cache_get_cache_and_entry(rte->relid, CACHE_FLAG_NONE, &hcache);
+			cagg = ts_continuous_agg_find_by_relid(rte->relid);
+
+			if (!ContinuousAggIsFinalized(cagg))
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("old format of continuous aggregate is not supported"),
+						 errhint("Run \"CALL cagg_migrate('%s.%s');\" to migrate to the new "
+								 "format.",
+								 NameStr(cagg->data.user_view_schema),
+								 NameStr(cagg->data.user_view_name))));
+			}
+
+			parent_mat_hypertable_id = cagg->data.mat_hypertable_id;
+			hcache = ts_hypertable_cache_pin();
+			ht = ts_hypertable_cache_get_entry_by_id(hcache, cagg->data.mat_hypertable_id);
+		}
 
 		if (TS_HYPERTABLE_IS_INTERNAL_COMPRESSION_TABLE(ht))
 			ereport(ERROR,
 					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 					 errmsg("hypertable is an internal compressed hypertable")));
 
-		/* there can only be one continuous aggregate per table */
-		switch (ts_continuous_agg_hypertable_status(ht->fd.id))
+		if (rte->relkind == RELKIND_RELATION)
 		{
-			case HypertableIsMaterialization:
-			case HypertableIsMaterializationAndRaw:
+			ContinuousAggHypertableStatus status = ts_continuous_agg_hypertable_status(ht->fd.id);
+
+			/* prevent create a CAGG over an existing materialization hypertable */
+			if (status == HypertableIsMaterialization ||
+				status == HypertableIsMaterializationAndRaw)
+			{
 				ereport(ERROR,
 						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						 errmsg("hypertable is a continuous aggregate materialization table")));
-			case HypertableIsRawTable:
-				break;
-			case HypertableIsNotContinuousAgg:
-				break;
-			default:
-				Assert(false);
+						 errmsg("hypertable is a continuous aggregate materialization table"),
+						 errhint("An hypertable or a continuous aggregate should be used instead "
+								 "of the materialization hypertable \"%s.%s\".",
+								 NameStr(ht->fd.schema_name),
+								 NameStr(ht->fd.table_name))));
```

Comment:
You can find the continuous aggregate for the materialization table, so I would suggest something along the lines:
```suggestion
						 errhint("Do you want to use continuous aggregate \"%s.%s\" instead?",
								 NameStr(schema_name),
								 NameStr(table_name))));
```
And then an `errdetail` with the materialization table name.